File: aoj/volume0/java/A0009.java
import java.util.Scanner;

class Main {
    public static void main(String args[]) {
        Scanner sc = new Scanner(System.in);

        boolean[] isPrimes = new boolean[1000000];
        for (int x = 2; x < 1000000 / 2 + 1; x++) {
            if (!isPrimes[x]) {
                for (int y = 2; y < 999999 / x + 1; y++) {
                    isPrimes[x * y] = true;
                }
            }
        }

        int count = 0;
        int[] nPrimes = new int[1000000];
        for (int i = 2; i < 1000000; i++) {
            if (!isPrimes[i]) {
                count++;
            }
            nPrimes[i] = count;
        }

        while (sc.hasNext()) {
            int n = sc.nextInt();
            System.out.println(nPrimes[n]);
        }
    }
}

File: aoj/volume0/ocaml/A0009.ml
let () =
    let max = 999999 in
    let is_primes = Array.make (max+1) true in
    for x = 2 to max / 2 do
        if is_primes.(x) then (
            for y = 2 to max / x do
                is_primes.(x*y) <- false
            done ;
        ) ;
    done ;

    let n_primes = Array.make (max+1) 0 in
    let count    = ref 0 in
    for x = 2 to max do
        if is_primes.(x) then (
            count := !count + 1
        ) ;
        n_primes.(x) <- !count
    done ;

    try
        while true do
            let n = read_int () in
            Printf.printf "%d\n" (n_primes.(n))
        done
    with
        End_of_file -> ()
